File: dotnet/base/Mcma.Client/AuthProvider.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mcma.Client
{
    public class AuthProvider : IAuthProvider
    {
        private Dictionary<string, Func<string, Task<IAuthenticator>>> RegisteredAuthTypes { get; }
            = new Dictionary<string, Func<string, Task<IAuthenticator>>>(StringComparer.OrdinalIgnoreCase);

        public IAuthProvider Add(string authType, Func<string, Task<IAuthenticator>> authenticatorFactory)
        {
            if (RegisteredAuthTypes.ContainsKey(authType))
                throw new Exception("Auth type '" + authType + "' has already been registered.");

            RegisteredAuthTypes[authType] = authenticatorFactory;

            return this;
        }

        public async Task<IAuthenticator> GetAsync(string authType, string authContext = null)
            => RegisteredAuthTypes.ContainsKey(authType) ? await RegisteredAuthTypes[authType](authContext) : null;
    }
}

File: nodejs/base/client/lib/auth-provider.js
class AuthProvider {
    constructor() {
        const registeredAuthTypes = {};

        this.add = (authType, authenticatorFactory) => {
            if (Object.keys(registeredAuthTypes).find(k => k.toLowerCase() === authType)) {
                throw new Error("Auth type '" + authType + "' has already been registered.");
            }
            if (typeof authenticatorFactory !== "function") {
                throw new Error("authenticatorFactory must be a function.");
            }

            registeredAuthTypes[authType] = authenticatorFactory;

            return this;
        };

        this.get = (authType, authContext) => {
            authType = Object.keys(registeredAuthTypes).find(k => k.toLowerCase() === (authType || "").toLowerCase());

            return authType && registeredAuthTypes[authType] && registeredAuthTypes[authType](authContext);
        };
    }
}

module.exports = {
    AuthProvider
};
